toolchain/lower/constant.cpp

```diff
@@ -0,0 +1,270 @@
+    CARBON_CHECK(const_id.is_template() && const_id.inst_id().index >= 0)
+        << "Unexpected constant ID " << const_id;
+    CARBON_CHECK(const_id.inst_id().index <= last_lowered_constant_index_)
+        << "Queried constant " << const_id << " that has not been lowered yet";
+    return constants_[const_id.inst_id().index];
+  }
+
+  // Returns a constant for the case of a value that should never be used.
+  auto GetUnusedConstant(SemIR::TypeId /*type_id*/) const -> llvm::Constant* {
+    // TODO: Consider using a poison value of the appropriate type.
+    return nullptr;
+  }
+
+  // Gets a callable's function. Returns nullptr for a builtin.
+  auto GetFunction(SemIR::FunctionId function_id) -> llvm::Function* {
+    return file_context_->GetFunction(function_id);
+  }
+
+  // Returns a lowered type for the given type_id.
+  auto GetType(SemIR::TypeId type_id) const -> llvm::Type* {
+    return file_context_->GetType(type_id);
+  }
+
+  // Returns a lowered value to use for a value of type `type`.
+  auto GetTypeAsValue() const -> llvm::Constant* {
+    return file_context_->GetTypeAsValue();
+  }
+
+  // Sets the index of the constant we most recently lowered. This is used to
+  // check we don't look at constants that we've not lowered yet.
+  auto SetLastLoweredConstantIndex(int32_t index) {
+    last_lowered_constant_index_ = index;
+  }
+
+  auto llvm_context() const -> llvm::LLVMContext& {
+    return file_context_->llvm_context();
+  }
+  auto llvm_module() const -> llvm::Module& {
+    return file_context_->llvm_module();
+  }
+  auto sem_ir() const -> const SemIR::File& { return file_context_->sem_ir(); }
+
+ private:
+  FileContext* file_context_;
+  llvm::MutableArrayRef<llvm::Constant*> constants_;
+  int32_t last_lowered_constant_index_ = -1;
+};
+
+// For each instruction kind that can produce a constant, there is a function
+// below to convert it to an `llvm::Constant*`:
+//
+// auto Emit<InstKind>AsConstant(ConstantContext& context,
+//                               SemIR::<InstKind> inst) -> llvm::Constant*;
+
+// For constants that are always of type `type`, produce the trivial runtime
+// representation of type `type`.
+#define CARBON_SEM_IR_INST_KIND_NOT_TYPE(...)
+#define CARBON_SEM_IR_INST_KIND_MAYBE_TYPE(...)
+#define CARBON_SEM_IR_INST_KIND_CONSTANT_SYMBOLIC_ONLY(...)
+#define CARBON_SEM_IR_INST_KIND(Name)                                      \
+  static auto Emit##Name##AsConstant(                                      \
+      ConstantContext& context, SemIR::Name /*inst*/) -> llvm::Constant* { \
+    return context.GetTypeAsValue();                                       \
+  }
+#include "toolchain/sem_ir/inst_kind.def"
+
+// Emits an aggregate constant of LLVM type `Type` whose elements are the
+// contents of `refs_id`.
+template <typename ConstantType, typename Type>
+static auto EmitAggregateConstant(ConstantContext& context,
+                                  SemIR::InstBlockId refs_id, Type* llvm_type)
```

Comment:
Maybe this is just me, but I'd lean towards these as member functions on ConstantContext rather than static functions. I'm not sure how much you're getting from the separation: it feels like it's actually increasing the code.

Alternately, given there's not file separation, maybe a big `switch` statement for a lot of these (rather than the function separation) would actually be okay? It might actually result in less code overall, considering duplication.